compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs

```diff
@@ -482,7 +482,32 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
                     fmt_printer(self, Namespace::TypeNS)
                         .comma_sep(generic_args.iter().copied().map(|arg| {
                             if arg.is_suggestable(self.tcx, true) {
-                                return arg;
+                                let ty_var_is_impl_trait = move |ty_vid| {
+                                    let mut infcx_inner = self.infcx.inner.borrow_mut();
+                                    let ty_vars = infcx_inner.type_variables();
+                                    let var_origin = ty_vars.var_origin(ty_vid);
+                                    if let TypeVariableOriginKind::TypeParameterDefinition(name, _) = var_origin.kind
+                                        && name.as_str().starts_with("impl ")
+                                    {
+                                        true
+                                    } else {
+                                        false
+                                    }
+                                };
```

Comment:
Is there a less hacky way to do this check?

Reply:
I think it's better to change `fn fmt_printer` to not use the name of`TypeVariableOriginKind::TypeParameterDefinition` if the `DefId` is from an argument position impl trait instead.

Reply:
I think reaching 11 indentation levels might be justification to refactor some of this to a function anyway 🙂

Reply:
> I think it's better to change `fn fmt_printer` to not use the name of`TypeVariableOriginKind::TypeParameterDefinition` if the `DefId` is from an argument position impl trait instead.

How do you check if a `DefId` is from an arg position `impl Trait`?

Reply:
`tcx.generics_of(tcx.parent(def_id))`, look at the https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/generics/struct.GenericParamDef.html for whether the type param is synthetic.

I am not 10% sure whether the `parent` is necessary 🤷 try it without, if it ices, add `tcx.parent`

Reply:
Thanks, I think that works, but I'm not sure if it's 100% correct.